### PR TITLE
Bump Go version to 1.17

### LINKED ISSUE #857: Bump go version to 1.17
**Describe the feature request**

Go 1.17 has been out for awhile now, and upstream k8s and others have already transitioned to using it. It would be good for us to come up with an official policy describing how we want to handle go updates. In the meantime, filing this so we can track some of the steps that we need to perform when we do decide to update.

- [ ] Update all `go.mod` files to include `go 1.17.x`
- [ ] Run `go mod tidy -go=1.17` to update our go.mod files to pick up [module graph](https://golang.org/doc/go1.17#tools) changes
- [x] Update our version of golangci-lint to address a [panic in one of the linters](https://github.com/golangci/golangci-lint/issues/2222)

Will add others if they come up.

**Describe alternatives you've considered**

We can stay on 1.16, but there are improvements in newer version that we would miss out on.

### PR DESCRIPTION
**Pull request by rajathagasthya**
### What this PR does / why we need it

Bumps Go version to 1.17.

### Which issue(s) this PR fixes

Related #857 

### Describe testing done for PR

```
$ make lint
$ go test ./...
$ make build-install-cli-all
```

### Release note

```release-note
Tanzu framwork now uses Go 1.17 for all builds and language feature support.
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Comment by stmcginnis** [comment]
Hmm, I hope this isn't something we're going to start hitting in our GitHub Actions regularly:

> Error: No space left on device

**Review by stmcginnis — APPROVED** [review]
Looks great to me. We should probably get some more feedback on the timing of moving to 1.17 though. My opinion: 1.17 has been out long enough now and has improvements that I think it would be good to make the bump.

**Review by vuil — APPROVED** [review]
lgtm. I second the bump as well. I had to build some targets in cluster-api main and it plain didn't work until I switched to 1.17.

File: addons/go.mod [hunk 1]
```diff
@@ -24,3 +24,60 @@ require (
 	sigs.k8s.io/cluster-api v1.0.1
 	sigs.k8s.io/controller-runtime v0.10.3
 )
+
+require (
```
Comment (ankeesler) [inline addons/go.mod:28]: where do these new `require` blocks come from? are they necessary?

**Review by ankeesler — APPROVED** [review]

File: addons/go.mod (hunk 1 above)
Comment (rajathagasthya) [reply addons/go.mod:28]: These are due to the new [module graph changes](https://go.dev/doc/go1.17#tools) in go 1.17. This is the result of running `go mod tidy -go=1.17` in each module.

**Comment by vijaykatam** [comment]
Do we need to bump base golang images in Dockerfiles as well?

**Comment by stmcginnis** [comment]
I do see a few other places where 1.16 is referenced. I haven't looked closely to see if these all require updating.

- `docs/cli/getting-started.md`
- `docs/release/release-notes-gathering-process.md`
- `addons/Dockerfile`
- `pkg/v1/tkr/Dockerfile`
- `addons/pinniped/post-deploy/Dockerfile`
- `pkg/v1/sdk/capabilities/Dockerfile`
- `pkg/v1/sdk/features/Dockerfile`
- `pkg/v1/builder/template/plugintemplates/gomod.tmpl`
- `pkg/v1/builder/template/plugintemplates/github_workflow_release.yaml.tmpl`
- `pkg/v1/builder/template/plugintemplates/gitlab-ci.yaml.tmpl`

**Comment by rajathagasthya** [comment]
@stmcginnis  I changed most of the files listed in your previous comment. The ones I didn't change were `plugintemplates` files. It [refers](https://github.com/vmware-tanzu/tanzu-framework/blob/e8a083109169b54f6d1ca1c5df16019dfcff1d8f/pkg/v1/builder/template/plugintemplates/gomod.tmpl#L9) to an old version of tanzu-framework, which I tried to update along with changing Go version to 1.17. But it doesn't seem to like it:

```
go list -m: github.com/vmware-tanzu/tanzu-framework@v0.15.0 requires
	sigs.k8s.io/cluster-api/test@v1.0.1 requires
	sigs.k8s.io/cluster-api@v0.0.0-00010101000000-000000000000: invalid version: unknown revision 000000000000
```

I think [this commit](https://github.com/vmware-tanzu/tanzu-framework/pull/1446/commits/aa4ae5ec4be5b5ed1ec73c48a8d43077e3726cb3) fixes it, but we will need to merge this PR first and then reference it in `plugintemplates/gomod.tmpl`. I will do it as a follow-up.

**Comment by stmcginnis** [comment]
Do we want to get any other folks giving explicit feedback? I think this is ready to go, but don't want to merge it if there are any other key stakeholders that should weigh in first.

**Comment by rajathagasthya** [comment]
@vuil I will leave it to you to add the `ok-to-merge` label. I think this is ready to go.

**Comment by stmcginnis** [comment]
Maybe a good argument for updating go versions, even if we're not super specific on the bugfix version: https://nvd.nist.gov/vuln/detail/CVE-2021-44716

**Comment by vuil** [comment]
Done.
May be a good idea to add a release note that stuff is built with 1.17 after the change

**Pull request merged**